storage: eliminate ambiguity for some Raft command retries

andreimatei opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12557)
<!-- Reviewable:end -->

andreimatei committed:
storage: eliminate ambiguity for some Raft command retries

Also add comments around reproposals.

andreimatei commented:
This is me trying to make sense of our code around reproposing and retrying (what I think used to be called "refurbishing" raft proposals).

The main thing I don't understand is how can we, in `Replica.refreshProposalsLocked()`, ask the caller to retry evaluating and sending a command, when it's ambiguous whether the original proposal was committed to the raft log or not (this is the `p.command.MaxLeaseIndex <= r.mu.state.LeaseAppliedIndex` case). The answer must be that Raft commands are generally idempotent, because of MVCC?
Because we do this retrying, I don't quite understand the point of `command.MaxLeaseIndex` / `replica.LeaseAppliedIndex` - it can't be to protect against "replays", as the comment on the field says (I've removed that word from the comment now). I think it must have something to do with wanting to guarantee ordering for the command application, but I don't quite see why that is necessary either, given that we have the command queue that also ensures some ordering.
Finally, I don't have a good feel about why we have the two categories of commands in `refreshPendingCommandsLocked` - the ones we repropose versus the ones we ask the proposer to retry. I superficially understand that the latter can't apply at the `MaxLeaseIndex` slot they've been assigned, but there's gotta be more to it.

If I get a better understanding of this stuff, I'd love to update the Proposer Evaluated KV RFC with it.

Thanks!

cc @petermattis

bdarnell commented:
Yes, we retry even in the ambiguous cases because we assume that no command can succeed twice unless it is idempotent. (but that's not 100% true: a KV Increment command outside of a transaction (or in a 1PC transaction whose Begin/EndTransaction commands have been optimized away) could succeed twice and produce inconsistent results. So we also rely on the fact that we don't currently do that, and we'll need to make some changes here before we can allow the SQL layer to optimize transactions like this). We retry because we believe it's common that in the ambiguous cases, the original command has been lost and the retry will succeed.

MaxLeaseIndex protects against raft-induced replays and reorderings, not the replays that we do ourselves via the loop in addWriteCmd.

There are two categories in refreshPendingCommandsLocked because raft reproposals are both cheaper and can never introduce ambiguity (thanks to MaxLeaseIndex). We believe this is a common case worth optimizing for. The more expensive reevaluated proposals are only done when we know that it's necessary because the previous command will never be applied by raft. 

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZeBBRr3MdeDGuV1x6z:-KZeBBRr3MdeDGuV1x7-:b-7sebwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/replica.go#L178)):*
> ```Go
> 	// does not succeed, care must be taken to correctly inform the caller via an
> 	// AmbiguousResultError.
> 	proposalShouldBeReevaluated
> ```

This is a good renaming; the old name might have made sense at one point but it definitely doesn't any more.

---

*[pkg/storage/replica.go, line 3005 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZe7eZc5WXH6tJYCbj-:-KZe7eZd8zUFt4LsF9Nh:blisacz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/replica.go#L3005)):*
> ```Go
> 		if p.command.MaxLeaseIndex <= r.mu.state.LeaseAppliedIndex {
> 			// The command's designated lease index slot was filled up and we don't
> 			// know if it was filled by this p.command or some other command.
> ```

But in most cases we do know - we've applied all commands up to this index, and `p` is still in `proposals`, so it must not have been applied. The one case where we're not sure is if we've caught up via a snapshot, so we haven't been watching each command as it applies. So I think we need to return a different ProposalRetry code depending on whether `reason == reasonSnapshotApplied`. 

---

*[pkg/storage/storagebase/proposer_kv.proto, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZe5OOmUwV5x-BxgKky:-KZe5OOmUwV5x-BxgKkz:bh2uxg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/storagebase/proposer_kv.proto#L139)):*
> ```protobuf
>   // counter has already reached (or exceeded) max_lease_index.
>   //
>   // The lease index is a reorder protection mechanism - we don't want Raft
> ```

It's replay protection too - this guarantees that a given command applies at most once, only if the ordering constraints hold, and it provides an upper bound after which we can be assured that the command will never execute. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-KZe9XTT72dw7QL_ebFu:b-bfkjyi)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Thanks Ben.

> MaxLeaseIndex protects against raft-induced replays and reorderings, not the replays that we do ourselves via the loop in addWriteCmd.

Well but what's the difference? Why can't we rely on idempotence/rejection when dealing with raft replays?

Could you detail the ordering considerations for raft proposals a bit? What constraints do we care about? When we blindly re-propose a command `p`, we know that (if it applies) it will apply before any command with a higher `MaxLeaseIndex`, but does that mean that it will apply before any command proposed after it was proposed? Can't it still apply after commands proposed by different nodes after `p` was proposed? Or are we relying on leases and the `OriginLease` check to prevent `p` from applying in this case? This would mean that the guarantee we need is that commands proposed by one proposer are never reordered. But then I'd again ask you if the ordering provided by the command queue is enough and if we could permit Raft to reorder commands.

If the command queue ordering is enough, then the next question is whether re-evaluation requests is ever needed; at the very least could we also directly re-propose the non-ambiguous cases that we currently ask the proposer to retry (i.e. the case you've talked about in the comment in `refreshProposalsLocked` - `p.command.MaxLeaseIndex <= r.mu.state.LeaseAppliedIndex` and `reason != reasonSnapshotApplied`)? Could we just assign a new `MaxLeaseIndex` and re-propose?


Sorry for rambling; the confusion is palpable, ain't it?

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZeBBRr3MdeDGuV1x6z:-K_0S8F-A1K4l1UlYiEq:b-yn305e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/replica.go#L178)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is a good renaming; the old name might have made sense at one point but it definitely doesn't any more.
</blockquote></details>

I've renamed it further to `proposalAmbiguousShouldBeReevaluated` to emphasize the ambiguous character, in contrast with `proposalIllegalLeaseIndex`.

---

*[pkg/storage/replica.go, line 3005 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZe7eZc5WXH6tJYCbj-:-K_0Qgaezxk3wF6EBFgn:bb7kxqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/replica.go#L3005)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

But in most cases we do know - we've applied all commands up to this index, and `p` is still in `proposals`, so it must not have been applied. The one case where we're not sure is if we've caught up via a snapshot, so we haven't been watching each command as it applies. So I think we need to return a different ProposalRetry code depending on whether `reason == reasonSnapshotApplied`. 
</blockquote></details>

I see. I'm now returning `proposalIllegalLeaseIndex` in that case. Does this seem good to you?

But please see PR questions about why/whether re-evaluation is needed at all.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZe5OOmUwV5x-BxgKky:-K_0R1MLZPYQgQ-h7PCM:brxwzjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f289c5436ab18ae597dd2347ce8b6eaa81c81cf/pkg/storage/storagebase/proposer_kv.proto#L139)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's replay protection too - this guarantees that a given command applies at most once, only if the ordering constraints hold, and it provides an upper bound after which we can be assured that the command will never execute. 
</blockquote></details>

ACK. I'll updated once I understand this a bit better :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_0K2IE2ZQ0tySGt1lp:bs4mo9m)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_Udax6fxVNMilLZc69:bq4ugx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
> ```Go
> 			// with the same MaxLeaseIndex) and is cheaper than asking the proposer to
> 			// re-evaluate and re-propose. Commands in this situation are frequent
> 			// enough.
> ```

Since you're in comment-mode, can you elaborate on when this situation occurs?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_a9WdmlzTZQ0Pjsgd5:b-jbw6qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Since you're in comment-mode, can you elaborate on when this situation occurs?
</blockquote></details>

I'm not entirely sure, tbh. I think it happens when Raft leadership changes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_bDG-r-UiVrQwHo3dy:bp87h1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm not entirely sure, tbh. I think it happens when Raft leadership changes?
</blockquote></details>

I'm trying to restore my understanding of this code. I believe we hit this code path when we proposed a command that never made it to the leader. That usually happens when we don't know who the leader is. When we do discover the leader we need to repropose the command. In local testing, by far the most common reason for these reproposals is the initial leader election after a split. Specifically, when a range is split when the next command is proposed to the RHS the leader is elected. After the election completes the command is reproposed for both `reasonNewLeader` and `reasonNewLeaderOrConfigChange`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
> > MaxLeaseIndex protects against raft-induced replays and reorderings, not the replays that we do ourselves via the loop in addWriteCmd.
> 
> Well but what's the difference? Why can't we rely on idempotence/rejection when dealing with raft replays?

There are two separate issues here: replays and reorderings. What we're calling raft-induced replays here are not actually happening in raft at all; they're caused by the `reproposals` loop in `refreshProposalsLocked`. This is essentially an optimization - we could instead return these commands to the loop in addWriteCommand to be fully retried (at the expense of calling `evaluateProposal` again and setting the `ambiguousResult` flag). 

Reorderings are a thornier problem - raft does not in general guarantee that commands will commit in the order in which they were proposed (although this is true when the proposals originate at the leader, so if we were willing to disallow all non-lease-related proposals when leader != leaseholder, we might be able to prevent reordering). If the command queue captures all relevant state, then we don't need to worry about reordering, but it doesn't (and even after #10413, we'll almost certainly end up with some special cases that are handled outside the command queue and will require other ways to prevent reordering). 

> Could you detail the ordering considerations for raft proposals a bit? What constraints do we care about? ... Or are we relying on leases and the OriginLease check to prevent p from applying in this case? This would mean that the guarantee we need is that commands proposed by one proposer are never reordered.

Yes, exactly. MaxLeaseIndex is tied to the OriginLease to prevent reordering of commands proposed by a single node, so that the leaseholder knows the order of all the commands that it proposes under its lease. This is what makes it possible for the command queue to not serialize all commands on the internal state of the range. 

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_blORkxGKIrVkdtUN-:-K_blORkxGKIrVkdtUN0:bw3al7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L173)):*
> ```Go
> 	// a Lease index it was not legal for. The command should be retried.
> 	proposalIllegalLeaseIndex
> 	// proposalShouldBeReevaluated indicates that it's ambiguous whether the command
> ```

Use the new name of the constant in this comment.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_c8VR-xoYaFQhMFV74:bmu5yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm trying to restore my understanding of this code. I believe we hit this code path when we proposed a command that never made it to the leader. That usually happens when we don't know who the leader is. When we do discover the leader we need to repropose the command. In local testing, by far the most common reason for these reproposals is the initial leader election after a split. Specifically, when a range is split when the next command is proposed to the RHS the leader is elected. After the election completes the command is reproposed for both `reasonNewLeader` and `reasonNewLeaderOrConfigChange`.
</blockquote></details>

Right. But we're not looking at the `reason` here, which doesn't seem right. I think the `proposedAtTicks` check above should only be done when `reason == reasonTicks`. For any other event we should be able to repropose regardless of how recently the previous attempt was sent. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_bsCpTuYlhX2kT1nyj:b-uyjgtl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_dgEzkDzBVQ_GV8Zul:bdf8ab6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Right. But we're not looking at the `reason` here, which doesn't seem right. I think the `proposedAtTicks` check above should only be done when `reason == reasonTicks`. For any other event we should be able to repropose regardless of how recently the previous attempt was sent. 
</blockquote></details>

We only pass a non-zero `refreshAtDelta` when `reason == reasonTicks`. If `refreshAtDelta == 0`, we'll never take the `proposedAtTicks` branch above. I wouldn't mind making this more explicit by adding a `reason == reasonTicks` check as well, but I don't think there is a problem with the existing code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_djLBkaijBuIa5-1jw:brgvafd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We only pass a non-zero `refreshAtDelta` when `reason == reasonTicks`. If `refreshAtDelta == 0`, we'll never take the `proposedAtTicks` branch above. I wouldn't mind making this more explicit by adding a `reason == reasonTicks` check as well, but I don't think there is a problem with the existing code.
</blockquote></details>

Ah, right. I was reading `refreshAtDelta` as a constant. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLj5noRKgpo_O2z4a:-K_eLj5noRKgpo_O2z4b:b-orltdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L192)):*
> ```Go
> 
> // proposalResult indicates the result of a proposal. Exactly one of
> // Reply, Err and ProposalRetry is set, and it represent the result of
> ```

represents

---

*[pkg/storage/replica.go, line 2990 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLqG_l34SaqO0abpd:-K_eLqG_l34SaqO0abpe:b-iug6du) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L2990)):*
> ```Go
> // refreshProposalsLocked goes through the pending proposals, notifying
> // proposers whose proposals need to be retried, and resubmitting proposals
> // which were likely dropped (but may still apply at a legal Lease index) - this
> ```

s/this way//

---

*[pkg/storage/replica.go, line 3004 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLxJznEFtBsCOptpO:-K_eLxJznEFtBsCOptpP:bm0wn7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3004)):*
> ```Go
> 	for idKey, p := range r.mu.proposals {
> 		if p.command.MaxLeaseIndex <= r.mu.state.LeaseAppliedIndex {
> 			// The command's designated lease index slot was filled up.  We got to
> ```

double space

---

*[pkg/storage/replica.go, line 3006 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eM5C3-wSP24ZdZXse:-K_eM5C3-wSP24ZdZXsf:bgvs3df) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3006)):*
> ```Go
> 			// The command's designated lease index slot was filled up.  We got to
> 			// LeaseAppliedIndex and p is still pending in r.mu.proposals; generally
> 			// this means that proposal p didn't commit, and it will sent back to the
> ```

will be?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
OK, so then it'd be fair to say that the `MaxLeaseIndex`'s role in preventing replays is just an optimization, correct? We could allow replays just fine, just like we allow retries. 
I would still prefer not to mention "preventing replays" in the documentation of `MaxLeaseIndex` because a) it's confusing why they'd be bad, given that we do retries and b) it suggests that it's important how we respond to them. In fact, it doesn't particularly matter how we respond - the `proposalIllegalLeaseIndex` that we return for them will not be processed by anybody (since nobody's listening for the response of a 2nd application). And thank god for that, otherwise, if somebody were to handle it just like we handle reorderings and perform a retry, that retry could get a non-ambiguous error when in fact it should have been ambiguous. 
You've said that reproposals "never introduce ambiguity ... thanks to MaxLeaseIndex" - but that's misleading; they don't introduce ambiguity because nobody cares about the result of a 2nd application of a command, not because we protect against replays. So the important thing is that we repropose with the same `CmdIDKey`; we could propose it with a different `MaxLeaseIndex` if we didn't care about reorderings.

Is this correct?

---

Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_blORkxGKIrVkdtUN-:-K_fvoOVWFHVHNazw0hW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L173)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Use the new name of the constant in this comment.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLj5noRKgpo_O2z4a:-K_fvwAXDR8Vhi33e3LB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

represents
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 2990 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLqG_l34SaqO0abpd:-K_fw-zq9jBpcjnELKoX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L2990)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/this way//
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 3004 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eLxJznEFtBsCOptpO:-K_fw1iWy7e4cMfGPnhg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3004)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

double space
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 3006 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_eM5C3-wSP24ZdZXse:-K_fw4JeXzVyUV8t9tQ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3006)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

will be?
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 3045 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_Udax5NhtqXTnC0mhx:-K_fz4T_X7f1_zy-Ypqz:b-lj1qru) ([raw file](https://github.com/cockroachdb/cockroach/blob/b33baa5219af2286ed57dd619497cb08895618ea/pkg/storage/replica.go#L3045)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Ah, right. I was reading `refreshAtDelta` as a constant. 
</blockquote></details>

Peter, I've pretty much pasted your explanation.

Ben I was also tripped up by the `refreshAtDelta`; I've made it more explicit and enforced that it only applies to `reasonTicks`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_fkNqCYLrmj-et54GK:btprgwg)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 3001 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_gEuL_R-6IQFOEU7Wu:-K_gEuL_R-6IQFOEU7Wv:b-5zrz9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/edfb66b802e6f96ee02a71a099bb107539c97b3d/pkg/storage/replica.go#L3001)):*
> ```Go
> func (r *Replica) refreshProposalsLocked(refreshAtDelta int, reason refreshRaftReason) {
> 	if refreshAtDelta != 0 && reason != reasonTicks {
> 		log.Fatal(context.TODO(), fmt.Sprintf(
> ```

log.Fatalf?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> OK, so then it'd be fair to say that the MaxLeaseIndex's role in preventing replays is just an optimization, correct? We could allow replays just fine, just like we allow retries.
> I would still prefer not to mention "preventing replays" in the documentation of MaxLeaseIndex because a) it's confusing why they'd be bad, given that we do retries and b) it suggests that it's important how we respond to them. In fact, it doesn't particularly matter how we respond - the proposalIllegalLeaseIndex that we return for them will not be processed by anybody (since nobody's listening for the response of a 2nd application). And thank god for that, otherwise, if somebody were to handle it just like we handle reorderings and perform a retry, that retry could get a non-ambiguous error when in fact it should have been ambiguous.

Correct.

> You've said that reproposals "never introduce ambiguity ... thanks to MaxLeaseIndex" - but that's misleading; they don't introduce ambiguity because nobody cares about the result of a 2nd application of a command, not because we protect against replays. So the important thing is that we repropose with the same CmdIDKey; we could propose it with a different MaxLeaseIndex if we didn't care about reorderings.

What I meant by that was that a failure caused by MaxLeaseIndex doesn't cause the ambiguousResult flag to be set in addWriteCommand. But I guess you're right that all that matters is that we return the first execution of a given CmdIDKey to the client. 

That reminds me that there's been talk about getting rid of CmdIDKey and identifying commands by their MaxLeaseIndex instead: 

	// TODO(andreimatei): idKey is legacy at this point: We could easily key
	// commands by their MaxLeaseIndex, and doing so should be ok with a stop-
	// the-world migration. However, various test facilities depend on the
	// command ID for e.g. replay protection.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_hkjlDlwRMoOIxzcQJ:bfewljq)*
<!-- Sent from Reviewable.io -->

tamird commented:
I have a WIP in which I've done that CmdIDKey -> MaxLeaseIndex change, but it doesn't quite work out because a ton of testing code (mostly written by @andreimatei and used for replay protection) uses CmdIDKey to uniquely identify commands and intercept or prevent replays in various places.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_jFrz9J2UeRtJP89_r:b-3ghvwr)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I've reworded the comment on reproposals to mention idKey too and explain how the 2nd result will be ignored.

Tamir CC me somehow on that WIP and maybe some Friday... We were talking about identifying commands in tests by MaxLeaseIndex+originLease instead of idKey.

Nice doing business with you fellas, I've learned things.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 3001 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_gEuL_R-6IQFOEU7Wu:-K_kY2AyxUaAGorc1FbD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/edfb66b802e6f96ee02a71a099bb107539c97b3d/pkg/storage/replica.go#L3001)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

log.Fatalf?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_kYdj4Q5S3tSTrNdg2:b-wphitu)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I've spent a fair bit of time understanding what `TestAmbiguousResultErrorOnReproposal` (renamed `TestAmbiguousResultErrorOnRetry`) is doing, as it was failing. I've tried to make it saner and document it a bit, but... I don't know if I got it. Something still feels wrong, either me or the test is a bit confused about what exactly it's testing.
@bdarnell if you could please take a look (`r5` vs `r4`) that'd be great. Or @spencerkimball if you happen to see this.
Thanks!

---

Review status: 2 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-K_pYv8cnPBxQGgoe8XM:b-2c39u6)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 2 of 6 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_test.go, line 6694 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v6US2HIasXO-xNWvf:-K_v6US2HIasXO-xNWvg:bmwdbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6694)):*
> ```Go
> 		// because of retries/reproposals or because of other writes in the same
> 		// batch. If we didn't do this, the retry that this test performs would
> 		// result in a WriteTooOldError.
> ```

Wow, [this logic](https://github.com/cockroachdb/cockroach/blob/99dca4ef8a42dc26dba53cb4218bc3e9ba73dc3c/pkg/storage/replica.go#L4075-L4089) is weird. I don't think that's what we want with a self-overlapping batch (in a txn, the last write to a given key wins. outside a txn, the *first* write does). 

---

*[pkg/storage/replica_test.go, line 6695 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v80Qf2k7mI-9k9t2t:-K_v80Qf2k7mI-9k9t2u:btx688g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6695)):*
> ```Go
> 		// batch. If we didn't do this, the retry that this test performs would
> 		// result in a WriteTooOldError.
> 		// NOTE(andrei): I don't quite understand what we're testing with the
> ```

It's just attempting to confirm the current behavior, that we don't return AmbiguousResultError for non-transactional writes. To that end, I think it would be better to put a single write in the batch and verify that we get the unambiguous WriteTooOldError instead of testing the surprising self-overlapping batch behavior. 

---

*[pkg/storage/replica_test.go, line 6698 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v8Lp86TGTXQkh3PiQ:-K_v8Lp86TGTXQkh3PiR:b1odv03) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6698)):*
> ```Go
> 		// retries of non-transactional batches. There's no "replay protection" for
> 		// non-transactional batches, but the results of their retries are
> 		// unreliable (you can easily get a WriteTooOldError).
> ```

Shouldn't you *always* get a WriteTooOldError? I'm not understanding how this test is passing today. The self-overlapping batch magic ignores the WriteTooOldError from the second write in the batch, but the first one should still be processed normally. 

---

*[pkg/storage/replica_test.go, line 6796 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_vCR2nNAnoupjqKyO_:-K_vCR2nNAnoupjqKyOa:b68x10k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6796)):*
> ```Go
> 		t.Run(c.name, func(t *testing.T) {
> 			// Install a proposal function which starts a retry before doing the
> 			// proposal.
> ```

Add "The batch will ultimately be executed twice, and the result of the second execution is returned to the client"

---

*[pkg/storage/replica_test.go, line 6810 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v9JbrWIZ5VHRiVTpt:-K_v9JbrWIZ5VHRiVTpu:b-tx4s3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6810)):*
> ```Go
> 					// command to be re-evaluated and re-proposed).
> 					origLeaseAppliedIndex := tc.repl.mu.state.LeaseAppliedIndex
> 					tc.repl.mu.state.LeaseAppliedIndex += 10
> ```

Do you still need to change LeaseAppliedIndex now that you're trying to generate retries instead of reproposals?

---

*[pkg/storage/replica_test.go, line 6821 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_vC-quSKoctkSxtvMX:-K_vC-quSKoctkSxtvMY:bzcxlto) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6821)):*
> ```Go
> 					// without blocking.
> 					tc.repl.mu.submitProposalFn = nil
> 					// We've just told the "client" to retry the request, but we still
> ```

"We've just told addWriteCmd to retry the request, so it will no longer listen for the result of this original proposal. However, we still want the original proposal to succeed (before the retry is processed), so submit it before releasing the lock. "

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 6 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 3110 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-Ka2goiDBvZWvP-pKMQR:-Ka2goiDBvZWvP-pKMQS:b6c1k6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica.go#L3110)):*
> ```Go
> 			// reproposing with the same MaxLeaseIndex and the same idKey as the
> 			// original (a possible second response for the same idKey would be
> 			// ignored).
> ```

Add a newline between paragraphs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: change retry detection mechanism for non-transactional batches

Before this patch, we had some dubious code in replica.executeBatch()
that did a best-effort retry detection based on whether a write ran into
another with the same timestamp. This was apparently intended to appease
some tests that did non-transactional batches and didn't like them being
applied more that once. Frankly, the code did not make much sense to me,
Ben and Spencer. Also, in the meantime we've gotten the MaxLeaseIndex
mechanism, which prevents double-application because of Raft
reproposals.
It was unclear how this code interacted with transactional batches, so
I've removed it. As a consequence, a retry of trasactional batch coming
after the transaction has ended could cause intents to be left around.
This only happens for retries, not reproposals, and moreover for
ambiguous retries.

As a result of this commit, non-transactional overlapping ranges are no
longer allowed.

andreimatei commented:
I've added a commit removing the weird replay protection for non-transactional batches. I'll probably squash it into one commit after review.
@bdarnell @spencerkimball PTAL when you can.

---

Review status: 2 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica_test.go, line 6694 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v6US2HIasXO-xNWvf:-Ka3YzJO_8BhWVfoC-EW:b-sjrayr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6694)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Wow, [this logic](https://github.com/cockroachdb/cockroach/blob/99dca4ef8a42dc26dba53cb4218bc3e9ba73dc3c/pkg/storage/replica.go#L4075-L4089) is weird. I don't think that's what we want with a self-overlapping batch (in a txn, the last write to a given key wins. outside a txn, the *first* write does). 
</blockquote></details>

In a non-txn batch, the last write still wins (it overwrites the key using an incremented timestamp, and *also* returns a `WriteTooOldError`). The code you pointed too only deals with whether or not a `WriteTooOldError` is returned to the client or not. The point of it is that we don't want  to return a `WTOE` when there's overlap, as the overlap explains the error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-Ka9ct7ewwRcSpcvXPpX:b59qqgl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 2 files at r3, 1 of 2 files at r4, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1916 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KaE2lm0RB8NLYJa5tLu:-KaE2lm0RB8NLYJa5tLv:b-81r3om) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L1916)):*
> ```Go
> 		}
> 		if pErr != nil {
> 			// If this isn't a non-transactional request or an end transaction with
> ```

This double negative is confusing. How about "If this is a transactional request other than an EndTransaction with commit=true, return the error immediately". 

---

*[pkg/storage/replica.go, line 4214 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KaE7_STqrrz8kGHkONn:-KaE7_SUywbMfs81mrPl:b6yw9dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L4214)):*
> ```Go
> 					if ba.Txn == nil {
> 						for _, union := range ba.Requests[:index] {
> 							if union.GetInner().Header().Overlaps(args.Header()) {
> ```

I was hoping we'd be able to get rid of the check for overlapping batches. Can't we now use a non-overlapping batch in the batch-of-puts test now? Where else do we use overlapping batches?

I don't love the new WillNotBeRetried flag. I think it's probably a win if it lets us get rid of this check, but not if it just adds to the subtle logic here. 

---

*[pkg/storage/replica_test.go, line 6796 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_vCR2nNAnoupjqKyO_:-KaE8MEKdXDIUxsDF7HV:b-boifx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6796)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add "The batch will ultimately be executed twice, and the result of the second execution is returned to the client"
</blockquote></details>

This comment and the next two are still unaddressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/roachpb/errors.proto:
```diff
@@ -184,8 +184,15 @@ message SendError {
 // An AmbiguousResultError indicates that an EndTransaction request
```
I think this entire comment needs to be revised. How about:
```
// An AmbiguousResultError indicates that a write request
// may have succeeded or failed, but the response was not
// received and the final result is ambiguous.
```
The other verbiage you added seems too specific and it's confusing to boot. This error is not just returned for non-transactional batches in the case you mention with `WriteTooOldError`s. It's also returned for all other errors which occur after a Raft re-proposal.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -170,11 +170,12 @@ const (
-	// so if the retry does not succeed, care must be taken to correctly inform
-	// the caller via an AmbiguousResultError.
-	proposalReproposed
+	// proposalAmbiguousShouldBeReevaluated indicates that it's ambiguous whether
```
I think this change is a step backwards. This is not clearer to me than the previous constant's naming and its associated comment.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -1904,17 +1905,18 @@ func (r *Replica) addWriteCmd(
-			// error without further ado.
-			if etArg, ok := ba.GetArg(roachpb.EndTransaction); !ok ||
-				!etArg.(*roachpb.EndTransactionRequest).Commit {
+			// If this isn't a non-transactional request or an end transaction with
```
s/isn't a non-transactional request/is a transactional request but doesn't include an end transaction/

s/`!(ok && !etArg.(*roachpb.EndTransactionRequest).Commit) && ba.Txn != nil`/`ba.Txn != nil && (!ok || !etArg.(*roachpb.EndTransactionRequest).Commit)`/

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -3055,21 +3064,59 @@ func (r *Replica) refreshProposalsLocked(refreshAtDelta int, reason refreshRaftR
+			// the proposer for a retry - the request needs to be re-evaluated and the
+			// command re-proposed with a new MaxLeaseIndex.
+			//
+			// Except in the case when we're refreshing because of
```
Awkward to start with "Except". Try a slight rephrase. Maybe "An exception is the case"

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -4147,50 +4195,50 @@ func (r *Replica) executeBatch(
 					ba.Txn.Timestamp.Forward(tErr.ActualTimestamp)
 					ba.Txn.WriteTooOld = true
 				} else {
+					// Note that, when the batch appears here to be non-transactional but in
```
s/Note that,/Note:/
s/appears here to be/appears to be/
/pushing the timestamp here/pushing the timestamp/

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -4147,50 +4195,50 @@ func (r *Replica) executeBatch(
-					// error unless there are no earlier commands in the batch
-					// which might have written the same key.
-					var overlap bool
+				// We got a WriteTooOldError. In case this is a transactional request,
```
:.(

I thought we were getting rid of this code block. Why do you think this is necessary? Your explanation looks to me like a premature optimization. Do we have some case in mind where this is crucial for performance? My intuition is the complexity here is a mistake given the avowed justification.

andreimatei commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 855 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-Ka9cmkGXf5y7uHtnlAg-r6-855:-KaOwnS7dL5IXweEiH6V:b-y44xv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/kv/txn_coord_sender.go#L855)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/are/is/
</blockquote></details>

code gone

---

*[pkg/roachpb/errors.proto, line 184 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_pYhcJwnRBiJLoyaLt-r6-184:-KaOxHUxLZ5wJ_lQOnsY:b9x1hql) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/roachpb/errors.proto#L184)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I think this entire comment needs to be revised. How about:
```
// An AmbiguousResultError indicates that a write request
// may have succeeded or failed, but the response was not
// received and the final result is ambiguous.
```
The other verbiage you added seems too specific and it's confusing to boot. This error is not just returned for non-transactional batches in the case you mention with `WriteTooOldError`s. It's also returned for all other errors which occur after a Raft re-proposal.
</blockquote></details>

The new comment is gone. 
I like that the original comment mentions `EndTransaction`, as that's the only command that can generate this in a transaction. I've just add "or a non-transactional write batch" to it.
It might still be worth mentioning somewhere that non-txn increments can be applied twice in case of retry, but I don't know where that should be specified.

---

*[pkg/storage/replica.go, line 3110 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-Ka2goiDBvZWvP-pKMQR:-KaP-7yGmSq-XL2JMKbj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica.go#L3110)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Add a newline between paragraphs.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 173 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-173:-KaEkxA4DvDQEQ8v-d77:b-4p8v9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L173)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I think this change is a step backwards. This is not clearer to me than the previous constant's naming and its associated comment.
</blockquote></details>

The name `proposalReproposed` is just misleading. Nothing has been reproposed.
It also doesn't suggest what needs to be done - re-evaluation and a retry, not a "reproposal".

---

*[pkg/storage/replica.go, line 1916 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-1916:-KaOuAO6P1jfF-LHUwwW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L1916)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/isn't a non-transactional request/is a transactional request but doesn't include an end transaction/

s/`!(ok && !etArg.(*roachpb.EndTransactionRequest).Commit) && ba.Txn != nil`/`ba.Txn != nil && (!ok || !etArg.(*roachpb.EndTransactionRequest).Commit)`/
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 1916 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KaE2lm0RB8NLYJa5tLu:-KaOvcqqtYL8X4pPiRiD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L1916)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This double negative is confusing. How about "If this is a transactional request other than an EndTransaction with commit=true, return the error immediately". 
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 3073 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-3073:-KaP-HAbVGJc1uHEJQpE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L3073)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Awkward to start with "Except". Try a slight rephrase. Maybe "An exception is the case"
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4198 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-4198:-KaOvjhJftQbwvuGxnGJ:b-upes59) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L4198)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

:.(

I thought we were getting rid of this code block. Why do you think this is necessary? Your explanation looks to me like a premature optimization. Do we have some case in mind where this is crucial for performance? My intuition is the complexity here is a mistake given the avowed justification.
</blockquote></details>

this wasn't about performance, it was about a test that was leaving an intent behind and then it was having trouble with that intent. But it was simpler than I thought to have the test do intent resolution, so now this code is gone.

---

*[pkg/storage/replica.go, line 4214 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KaE7_STqrrz8kGHkONn:-KaOw3A9TOn3hOHwKfO5:bfdflog) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L4214)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I was hoping we'd be able to get rid of the check for overlapping batches. Can't we now use a non-overlapping batch in the batch-of-puts test now? Where else do we use overlapping batches?

I don't love the new WillNotBeRetried flag. I think it's probably a win if it lets us get rid of this check, but not if it just adds to the subtle logic here. 
</blockquote></details>

ok, non-txn overlapping batches are now disallowed. Only two tests seemed to be using them and I've fixed them.

---

*[pkg/storage/replica.go, line 4232 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-4232:-KaOwds5LjQh9eS9jwRq:b-y44xv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L4232)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

s/Note that,/Note:/
s/appears here to be/appears to be/
/pushing the timestamp here/pushing the timestamp/
</blockquote></details>

code gone

---

*[pkg/storage/replica_test.go, line 6695 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v80Qf2k7mI-9k9t2t:-KaP-OJZj3c3YUwD3hw4:b7q4043) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6695)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's just attempting to confirm the current behavior, that we don't return AmbiguousResultError for non-transactional writes. To that end, I think it would be better to put a single write in the batch and verify that we get the unambiguous WriteTooOldError instead of testing the surprising self-overlapping batch behavior. 
</blockquote></details>

done in the next commit

---

*[pkg/storage/replica_test.go, line 6698 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v8Lp86TGTXQkh3PiQ:-KaP-gZTSh2zlf3W0Jaj:brktm20) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6698)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Shouldn't you *always* get a WriteTooOldError? I'm not understanding how this test is passing today. The self-overlapping batch magic ignores the WriteTooOldError from the second write in the batch, but the first one should still be processed normally. 
</blockquote></details>

as we discussed, the intentions of the code in `executeBatch` were hard to decipher. The `WriteTooOldError` was just swallowed.

---

*[pkg/storage/replica_test.go, line 6796 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_vCR2nNAnoupjqKyO_:-KaOyu5qQylPgQQVJYJt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6796)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This comment and the next two are still unaddressed.
</blockquote></details>

Done.

---

*[pkg/storage/replica_test.go, line 6810 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_v9JbrWIZ5VHRiVTpt:-KaOzTRAIuV06_1Q6Rw8:b-4rly4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6810)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Do you still need to change LeaseAppliedIndex now that you're trying to generate retries instead of reproposals?
</blockquote></details>

the fact that we're increasing the LeaseAppliedIndex is what causes a retry and not a reproposal.

---

*[pkg/storage/replica_test.go, line 6821 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-K_vC-quSKoctkSxtvMX:-KaOzyIaMin1YrZCA0kc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f35f965c065623104f0b4113863903c30dca318/pkg/storage/replica_test.go#L6821)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"We've just told addWriteCmd to retry the request, so it will no longer listen for the result of this original proposal. However, we still want the original proposal to succeed (before the retry is processed), so submit it before releasing the lock. "
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 7 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 173 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-KZcSqb8HrDGc8R9B_g1-r6-173:-KaSjy8W2m0Cwt39ET02:btjlth7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11a66fc41f5682574b94e7bb7f6097ea938a07a/pkg/storage/replica.go#L173)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The name `proposalReproposed` is just misleading. Nothing has been reproposed.
It also doesn't suggest what needs to be done - re-evaluation and a retry, not a "reproposal".
</blockquote></details>

Yeah, `proposalReproposed` didn't make any sense (especially the use of the past tense). As this code is now used, it could be renamed to `proposalSnapshotApplied` to be more clear about the nature of the ambiguity. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12557#-:-KaSnu92wmJJ6YDfJtQT:b-uwkowd)*
<!-- Sent from Reviewable.io -->
